lnrpc/routerrpc/router_server_test.go

```diff
@@ -214,3 +217,170 @@ func TestTrackPaymentsNoInflightUpdates(t *testing.T) {
+	return &streamHtlcEventsMock{
+		ctx:            ctx,
+		sentFromServer: make(chan *HtlcEvent, 20),
+	}
+}
+
+// TestSubscribeHtlcEventsReturnsOnCancelContext tests whether
+// SubscribeHtlcEvents returns when the stream context is cancelled.
+func TestSubscribeHtlcEventsReturnsOnCancelContext(t *testing.T) {
+	t.Parallel()
+	request := &SubscribeHtlcEventsRequest{}
+
+	streamCtx, cancelStream := context.WithCancel(context.Background())
+	stream := makeStreamHtlcEventsMock(streamCtx)
+
+	// Create server mock for htlc propagation.
+	htlcServerMock := subscribe.NewServer()
+	err := htlcServerMock.Start()
+	if err != nil {
+		t.Fatalf("starting htlc server failed with: %s", err)
+	}
+	defer func() {
+		err := htlcServerMock.Stop()
+		if err != nil {
+			t.Fatalf("failed stopping htlc server: %s", err)
+		}
+	}()
+
+	server := &Server{
+		cfg: &Config{
+			RouterBackend: &RouterBackend{
+				SubscribeHtlcEvents: func() (
+					*subscribe.Client, error) {
+
+					return htlcServerMock.Subscribe()
+				},
+			},
+		},
+	}
+
+	// Cancel stream immediately.
+	cancelStream()
+
+	// Make sure the call returns.
+	err = server.SubscribeHtlcEvents(request, stream)
+	require.Equal(t, context.Canceled, err)
+}
+
+// TestHtlcEvents tests whether all htlc events are propagated to the client.
+func TestSubscribeHtlcEvents(t *testing.T) {
+	t.Parallel()
+	request := &SubscribeHtlcEventsRequest{}
+
+	// Create server mock for htlc propagation.
+	htlcServerMock := subscribe.NewServer()
+	err := htlcServerMock.Start()
+	if err != nil {
+		t.Fatalf("starting htlc server failed with: %s", err)
+	}
+	defer func() {
+		err := htlcServerMock.Stop()
+		if err != nil {
+			t.Fatalf("failed stopping htlc server: %s", err)
+		}
+	}()
+
+	streamCtx, cancelStream := context.WithCancel(context.Background())
+	stream := makeStreamHtlcEventsMock(streamCtx)
+	defer cancelStream()
+
+	server := &Server{
+		cfg: &Config{
+			RouterBackend: &RouterBackend{
+				SubscribeHtlcEvents: func() (
+					*subscribe.Client, error) {
+
+					return htlcServerMock.Subscribe()
+				},
+			},
+		},
+	}
+
+	// Listen to htlc events in a goroutine.
+	go func() {
+		err := server.SubscribeHtlcEvents(request, stream)
+		require.Equal(t, context.Canceled, err)
+	}()
+
+	// Make sure the client is subscribed before we send htlc events.
+	time.Sleep(50 * time.Millisecond)
+
+	// Send all types of HTLC events through the server and see whether
+	// the client receives the right types in right order.
+	err = htlcServerMock.SendUpdate(&htlcswitch.ForwardingFailEvent{})
+	if err != nil {
+		t.Fatalf("failed sending htlc event to server: %s", err)
+	}
+	err = htlcServerMock.SendUpdate(&htlcswitch.SettleEvent{})
+	if err != nil {
+		t.Fatalf("failed sending htlc event to server: %s", err)
+	}
+	err = htlcServerMock.SendUpdate(&htlcswitch.FinalHtlcEvent{})
+	if err != nil {
+		t.Fatalf("failed sending htlc event to server: %s", err)
+	}
+	// Make sure the LinkError is populated otherwise the test fails.
```

Comment:
The comment speaks for itself, though this is the part where the test will fail if we just pass the empty LinkFailEvent